core/include/scipp/core/transform_common.h

```diff
@@ -78,32 +78,47 @@ template <typename Op> assign_unary(Op) -> assign_unary<Op>;
 /// actually called since flag presence is checked via the base class of the
 /// operator.
 namespace transform_flags {
+
+/// Base and NULL flag. Do not test for this type.
+struct Flag {
+  void operator()() const {};
+};
+
+/// Helper to conditionally apply a given flag (T) under condition (B) otherwise
+/// Null Flag
+template <bool B, typename T>
+using enable_flag_if = std::conditional_t<B, T, Flag>;
+
+namespace {
+
+struct no_out_variance_t : Flag {};
 /// Add this to overloaded operator to indicate that the operation does not
 /// return data with variances, regardless of whether inputs have variances.
-static constexpr auto no_out_variance = []() {};
-using no_out_variance_t = decltype(no_out_variance);
+constexpr auto no_out_variance = no_out_variance_t{};
 
+template <int N> struct expect_no_variance_arg_t : Flag {};
 /// Add this to overloaded operator to indicate that the operation does not
 /// support variances in the specified argument.
-template <int N> static constexpr auto expect_no_variance_arg = []() {};
 template <int N>
-using expect_no_variance_arg_t = decltype(expect_no_variance_arg<N>);
+constexpr auto expect_no_variance_arg = expect_no_variance_arg_t<N>{};
 
+template <int N> struct expect_variance_arg_t : Flag {};
 /// Add this to overloaded operator to indicate that the operation requires
 /// variances in the specified argument.
-template <int N> static constexpr auto expect_variance_arg = []() {};
-template <int N> using expect_variance_arg_t = decltype(expect_variance_arg<N>);
+template <int N>
+constexpr auto expect_variance_arg = expect_variance_arg_t<N>{};
 
+struct expect_in_variance_if_out_variance_t : Flag {};
 /// Add this to overloaded operator to indicate that the in-place operation
 /// requires inputs to have a variance if the output has a variance.
-static constexpr auto expect_in_variance_if_out_variance = []() {};
-using expect_in_variance_if_out_variance_t =
-    decltype(expect_in_variance_if_out_variance);
+constexpr auto expect_in_variance_if_out_variance =
+    expect_in_variance_if_out_variance_t{};
 
-static constexpr auto expect_all_or_none_have_variance = []() {};
-using expect_all_or_none_have_variance_t =
-    decltype(expect_all_or_none_have_variance);
+struct expect_all_or_none_have_variance_t : Flag {};
+static constexpr auto expect_all_or_none_have_variance =
+    expect_all_or_none_have_variance_t{};
 
+} // namespace
```

Comment:
Why the anonymous namespace? I see you define the instances as `static` in some cases, but `static` is missing in others, is that why you needed it?

Reply:
Just an incomplete refactor. Should just need one or the other for internal linkage. I was moving to anonymous namespace.

Reply:
So this file is included in dozens of other files. Isn't `static` the way to go, instead of the anonymous namespace? Putting the classes into this namespace effectively means that every file including this one has *different* flags? The may only be internal so it does not matter in practice, it it feels odd.